Cross-browser shield text

ZeLonewolf opened the pull request:
Fixes #625 

This PR fixes differences between chrome and webkit by using fudge factors.

It also fixes several "I don't know how this even worked before" bugs in the banner layout code, and does some related typescript cleanup in the affected routines.

They are not pixel-perfect between browsers but they're very close.  Tested on Ubuntu.

**Webkit**
![image](https://github.com/ZeLonewolf/openstreetmap-americana/assets/3254090/8fa45291-fc1f-40e6-93c6-59a5528287d4)

**Chrome**
![image](https://github.com/ZeLonewolf/openstreetmap-americana/assets/3254090/431945a0-99f7-4f9e-b8e8-251c58812c50)

**Firefox**
![image](https://github.com/ZeLonewolf/openstreetmap-americana/assets/3254090/0037a457-d758-4724-84f7-72795b5fab72)

ZeLonewolf commented on shieldlib/src/shield_banner.ts:
```diff
@@ -133,7 +133,7 @@ function drawBannerTextComponent(
 ): void {
   const bannerPadding = {
-    top: r.options.bannerPadding,
+    top: 0,
```
Removed because the padding is accomplished by placing a gap between successive layouts, not by buffering the layout itself.

ZeLonewolf commented on shieldlib/src/shield_banner.ts:
```diff
@@ -161,7 +161,7 @@ function drawBannerTextComponent(
       textLayout.xBaseline,
       textLayout.yBaseline +
-        bannerIndex * r.px(r.options.bannerHeight - r.options.bannerPadding)
+        bannerIndex * r.px(r.options.bannerHeight + r.options.bannerPadding)
```
I'm not even sure how this code worked with this bug.

ZeLonewolf commented on shieldlib/src/shield_text.ts:
```diff
@@ -157,34 +165,49 @@ export function layoutShieldText(
-  ctx.textAlign = "center";
+  ctx.textAlign = "left";
   ctx.textBaseline = "top";
 
-  var metrics = ctx.measureText(text);
+  let metrics: TextMetrics = ctx.measureText(text);
 
-  var textWidth = metrics.width;
-  var textHeight = metrics.actualBoundingBoxDescent;
+  let textWidth: number =
+    Math.abs(metrics.actualBoundingBoxLeft) +
+    Math.abs(metrics.actualBoundingBoxRight);
+  let textHeight: number =
+    Math.abs(metrics.actualBoundingBoxDescent) +
+    Math.abs(metrics.actualBoundingBoxAscent);
 
-  var availHeight = bounds.height - padTop - padBot;
-  var availWidth = bounds.width - padLeft - padRight;
+  //Adjust for excess descender text height across browsers
+  textHeight *= 0.9;
 
-  var xBaseline = padLeft + availWidth / 2;
+  //Adjust for excess text height measured in Webkit engine specifically
+  if (isRunningInWebKit()) {
+    textHeight *= 0.54;
+  }
+
+  let availHeight: number = bounds.height - padTop - padBot;
+  let availWidth: number = bounds.width - padLeft - padRight;
+
+  let xBaseline: number = padLeft + availWidth / 2;
```
This is the main fix.  We check both ascent/descent and left/right to get height/width, and then we make empirical adjustments to account for how the different browsers actually handle these fonts.

1ec5 replied:
This makes me wonder if WebKit is reporting the full line height as the computer text height.

ZeLonewolf commented on shieldlib/src/shield_text.ts:
```diff
@@ -157,34 +165,49 @@ export function layoutShieldText(
+    Math.abs(metrics.actualBoundingBoxAscent);
 
-  var availHeight = bounds.height - padTop - padBot;
-  var availWidth = bounds.width - padLeft - padRight;
+  //Adjust for excess descender text height across browsers
+  textHeight *= 0.9;
 
-  var xBaseline = padLeft + availWidth / 2;
+  //Adjust for excess text height measured in Webkit engine specifically
+  if (isRunningInWebKit()) {
+    textHeight *= 0.54;
+  }
+
+  let availHeight: number = bounds.height - padTop - padBot;
+  let availWidth: number = bounds.width - padLeft - padRight;
+
+  let xBaseline: number = padLeft + availWidth / 2;
 
   let textLayoutFunc = drawTextFunctions[textLayoutDef.constraintFunc];
 
-  let textConstraint = textLayoutFunc(
-    { height: availHeight, width: availWidth },
-    { height: textHeight, width: textWidth },
+  let spaceAvail: Dimension = { height: availHeight, width: availWidth };
+  let measuredTextBounds: Dimension = { height: textHeight, width: textWidth };
+
+  let textConstraint: TextTransform = textLayoutFunc(
+    spaceAvail,
+    measuredTextBounds,
```
Just a nicety to add typescript typing and make it clear what's happening.

ZeLonewolf commented on shieldlib/src/shield_text.ts:
```diff
@@ -195,20 +218,23 @@ export function layoutShieldText(
 
   ctx.font = Gfx.shieldFont(fontSize, r.options.shieldFont);
-  ctx.textAlign = "center";
+  ctx.textAlign = "left";
   ctx.textBaseline = "top";
```
top/left makes the positioning math a bit easier and more consistent.

zekefarwell commented:
I tested this out on Mac OS 13, Windows 11, iOS 17, and Android 14.  Looks quite consistent across all the browsers I tested!  Examples on the interstate shield only for a focused comparison:

![Windows-Chrome](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/841cac73-5e92-4fff-a8ff-ff806d9144b7)
Chrome on Windows

![Mac-Chrome](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/8f0d0b22-4a03-4349-ae2f-25f74cc461c3)
Chrome on Mac

![Android-Chrome](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/dc816c7d-6983-4889-a0be-27d1a1dac9a0)
Chrome on Android

![Mac-Safari](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/f8a1d4c5-4c5f-44cc-9ece-076bb8a9b463)
Safari on Mac

![iOS-Safari](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/8a6793cb-b90f-40e9-9696-6f5566288e30)
Safari on iOS

![Windows-FF](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/0ba37c98-baa3-4581-8905-d9cbea349861)
Firefox on Windows

![Mac-FF](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/5e650187-90f6-4f9e-a164-effa2ce6c75f)
Firefox on Mac

![Android-FF](https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/4a87eeb8-4570-40be-aab5-2576fe7079ff)
Firefox on Android

Definitely some vertical padding adjust needed for this and other shields, but that can probably be a follow up PR.  Another issue I'm seeing is Chrome on Windows getting the wide shield blank for 2 digit refs.  On the plus side, Safari and Firefox on Windows get the narrow shield blank for 111.

ZeLonewolf commented:
> Definitely some vertical padding adjustment needed for this and other shields, but that can probably be a follow up PR. Another issue I'm seeing is Chrome on Windows getting the wide shield blank for 2 digit refs. On the plus side, Safari and Firefox on Windows get the narrow shield blank for 111.

It looks like that only happens for the slightly wider 2-digit numbers, so I suspect I can just micro-adjust that threshold and achieve the intended effect.

ZeLonewolf commented:
Reducing the threshold from 12 to 11.8 seems to be enough for the right 2-digit to 3-digit transition on Windows/Chrome.

zekefarwell commented:
> Reducing the threshold from 12 to 11.8 seems to be enough for the right 2-digit to 3-digit transition on Windows/Chrome

Looks like it.  I can no longer reproduce.  👍

I noticed a couple of regressions that it may may be nice to resolve before merging this.  The banner text is larger on this branch than on main.  Also in some cases the banner text size changes between the narrow and wide variants of a certain shield.

<img width="649" alt="banner-text-PR-1010" src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/155ec44c-1161-4e1d-86d9-7eaddb0836ba">

This PR


<img width="649" alt="banner-text-main" src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/06d59d48-bf8b-4802-99ad-1e61d42639cf">

Main branch


Some shields have a thicker top border on this branch that isn't present on main.  I'm only seeing this issue on (some) bannered shields so it may be related to the banner text somehow.

<img width="648" alt="thick-border-PR-1010" src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/0251197c-155a-4edc-bae8-ddb5a7118f57">

This PR


<img width="648" alt="thick-border-main" src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/281482/f69ce2f8-c5ce-4a9b-8977-087e93adef57">

Main Branch

claysmalley commented:
> Some shields have a thicker top border on this branch that isn't present on main. I'm only seeing this issue on (some) bannered shields so it may be related to the banner text somehow.

This appears to be affecting shields generated by draw functions. SVG shields look okay.

ZeLonewolf commented:
[755117b](https://github.com/ZeLonewolf/openstreetmap-americana/pull/1010/commits/755117b981b481da10895803230ac55b5db76ca9) fixes the weirdness we were seeing with some shields.  I simplified the code that draws shields when banners are present and removed an extra draw that wasn't supposed to be there.

1ec5 reviewed (approved):
Looks better than before in MobileSafari:

<img width="667" alt="Safari" src="https://github.com/ZeLonewolf/openstreetmap-americana/assets/1231218/31f475ab-24ce-45d2-a93c-29122a533017">
